### PR TITLE
8332697: ubsan: shenandoahSimpleBitMap.inline.hpp:68:23: runtime error: signed integer overflow: -9223372036854775808 - 1 cannot be represented in type 'long int'

### PR DESCRIPTION
**Pull request by earthling-amzn**
Use a template version of `right_n_bits` to use the same type for minuend and subtrahend.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332697](https://bugs.openjdk.org/browse/JDK-8332697): ubsan: shenandoahSimpleBitMap.inline.hpp:68:23: runtime error: signed integer overflow: -9223372036854775808 - 1 cannot be represented in type 'long int' (**Bug** - P4)

### Reviewers
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**)
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Author)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21236/head:pull/21236` \
`$ git checkout pull/21236`

Update a local copy of the PR: \
`$ git checkout pull/21236` \
`$ git pull https://git.openjdk.org/jdk.git pull/21236/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21236`

View PR using the GUI difftool: \
`$ git pr show -t 21236`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21236.diff">https://git.openjdk.org/jdk/pull/21236.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21236#issuecomment-2380088556)

### CODE REVIEW

**Review by phohensee — COMMENTED** [review]
Rather than define a new method get_right_n_bits(), why not just replace the definition of right_n_bits() in globalDefinitions.hpp? The C++ compiler will inline and optimize both.

**Comment by earthling-amzn** [comment]
I tried that, but there is a warning at the macro declaration:
```
// (note: #define used only so that they can be used in enum constant definitions)
#define nth_bit(n)        (((n) >= BitsPerWord) ? 0 : (OneBit << (n)))
#define right_n_bits(n)   (nth_bit(n) - 1)
```
There are many usages of `right_n_bits` that use an unnamed enum constant, which will not accept a cast from a numeric type.

**Comment by phohensee** [comment]
In that case, I'd put get_right_n_bits() in globalDefinitions.hpp because it's generally useful, and a comment on why, of course. :)

**Comment by earthling-amzn** [comment]
I'd prefer to keep the scope of these changes limited to this file. This is the only place where this seems to be a problem in the JDK and I think it's just because we abused a macro we shouldn't have.

**Review by phohensee — APPROVED** [review]
Not using right_n_bits() at works for me.

**Review by kdnilsen — APPROVED** [review]
Thanks.

File: src/hotspot/share/gc/shenandoah/shenandoahSimpleBitMap.inline.hpp [hunk 1]
```diff
@@ -27,15 +27,22 @@
 
 #include "gc/shenandoah/shenandoahSimpleBitMap.hpp"
 
+inline uintx tail_mask(uintx bit_number) {
+  if (bit_number >= BitsPerWord) {
+    return -1;
+  }
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahSimpleBitMap.inline.hpp:33]: When would we call here with `bit_number >= BitsPerWord` ? If never, may be we assert that?

Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahSimpleBitMap.inline.hpp:33]: It's called with `bit_number == BitsPerWord` (I tried an assertion first).

**Pull request closed without merging**